File: perl/pe008.pl
#!/usr/bin/perl

# Project Euler problem 8
# by Taylor Fausak

# Finds the greatest product of five consecutive digits in a number. Simply
# makes the number a string and moves through it.

use strict;
use warnings;

my $length = 5;
my $max = -1;
my $n = '';

open(INPUT, '<', $ARGV[0]) or die $!;

while (my $line = <INPUT>) {
    $line =~ s/^\s*(\d+)\s*$/$1/; # trim whitespace
    $n .= $line;
}

close(INPUT) or die $!;

for (my $offset = 0; $offset < length($n) - $length; $offset++) {
    my $product = 1;
    
    foreach my $digit (split('', substr($n, $offset, $length))) {
        $product *= $digit;
    }
    
    if ($product > $max) {
        $max = $product;
    }
}

print $max . "\n";

File: ruby/pe008.rb
#!/usr/bin/env ruby
# http://projecteuler.net/problem=8

number = File.read(ARGV[0]).inject { |number, line| number += line.chomp }
max = 0
(0..number.length - 5).each do |offset|
  digits = number[offset..offset + 4].chars.collect { |digit| digit.to_i }
  product = digits.inject { |product, digit| product *= digit }
  max = product if product > max
end
puts max
